### ISSUE TITLE
ONNX export for quantization?

### ISSUE THREAD

**Issue by rotx-maxim** [NervanaSystems/distiller#23]
I am experimenting with exporting a quantized network to ONNX. This ultimately does not succeed because there is no `round` operator in ONNX, and PyTorch does not define an `ATen` for round either.

I'm not sure what the best strategy would be (perhaps using `floor`, which exists in ONNX but is still missing in the PyTorch exporter), and some guidance would be appreciated.

In case anyone would like to duplicate the experiment, the first step was to modify the `forward()` method in `ClippedLinearQuantization`. Instead of the call to `LinearQuantizeSTE.apply` (the PyTorch ONNX exporter doesn't know what to do with that), inline the contents of `LinearQuantizeSTE.forward`, like so:

        def forward(self, input_):
            input_ = clamp(input_, 0, self.clip_val, self.inplace)
            if self.inplace:
                input_.mark_dirty(input_)
            input_ = linear_quantize(input_, self.scale_factor, self.inplace)
            if self.dequantize:
                input_ = linear_dequantize(self.input_, self.scale_factor, self.inplace)
            return input_

This should be functionally equivalent and the export trace will now complain about `round`.  In `q_utils.py`, modify `linear_quantize` (for example, remove the calls to `round_()` and `round()` and replace them with... something else).

**Comment by vinutah** [comment]
Hi Robert, I am experimenting with the same thing. 
1. There is an enhancement request on `onnx`  `quantized tensors #376`
    https://github.com/onnx/onnx/issues/376 which is Open
2. But have you checkout out `Extensibility of ONNX operators #168` https://github.com/onnx/onnx/issues/168.

If you figure out a neat way of exporting the quantized model out as .`onnx` please do let me know too.

**Comment by guyjacob** [comment]
@vinutah: Thanks for the great pointers. You'll likely find this [discussion group](https://gitter.im/onnx/quantization) relevant as well. And, [onnx/onnx#809](https://github.com/onnx/onnx/pull/809) is another quantization related PR, not sure if it's instead of the one you linked to above or in addition to it.

From our side, ONNX export of quantized models is something we haven't tackled yet. ONNX support for INT data types and operations is a necessary first step. Then, it might be possible to extend the PyTorch exporter even without native PyTorch support for low-precision (see this [ONNX-PyTorch tutorial](https://github.com/onnx/tutorials/blob/master/tutorials/PytorchAddExportSupport.md)). Note that some sort of quantization support is expected in PyTorch 1.0, which will probably be a big help.

@rotx-maxim: A couple of thoughts on your specific ideas above:

1. I'm not too familiar with ATen, but as far as I can see there it does expose a `round` function - see [here](https://github.com/pytorch/pytorch/blob/master/aten/doc/Tensor.h) and [here](https://github.com/pytorch/pytorch/blob/master/aten/doc/Functions.h).
2. Not sure if that's what you meant, but just replacing `round` with `floor` will likely result in significant accuracy loss due to the bias it'll introduce to the quantization. Doing `floor(tensor + 0.5)` is a better idea.
3. I assume you're referring to schemes that require training with quantization (e.g. DoReFa or WRPN that we currently have implemented). In these cases, inlining the contents of `linear_quantize` as you propose is functionally equivalent **only for the forward pass**. In the backward pass we don't use the derivative of the `round` op as-is - we replace it with a "straight-through estimator", or STE (see details [here](https://nervanasystems.github.io/distiller/quantization/index.html#training-with-quantization)). Hence, we have to use a dedicated `autograd.Function` for the quantization operation, so we can override the backward function. Inlining will cause the derivative of `round` to be used as-is, which will likely result in gradients of 0 almost everywhere, and the model will not learn.

**Comment by Aeroxander** [comment]
Would be awesome if there was a similar feature as when you convert a Tensorflow model to tf.js with a quantization flag option!

**Comment by guyjacob** [comment]
I forgot that this issue was open :)
Opened #255 in light of the official release of quantization support in ONNX. Will continue to track from there.
